src/group_theory/bundled_subgroup.lean

```diff
@@ -0,0 +1,700 @@
+membership of a subgroup's underlying set.
+
+## Tags
+subgroup, subgroups
+-/
+
+variables {G : Type*} [group G]
+variables {A : Type*} [add_group A]
+
+/-- A subgroup of a group `G` is a subset containing 1, closed under multiplication
+and closed under multiplicative inverse. -/
+structure subgroup (G : Type*) [group G] extends submonoid G :=
+(inv_mem' {x} : x ∈ carrier → x⁻¹ ∈ carrier)
+
+/-- An additive subgroup of an additive group `G` is a subset containing 0, closed
+under addition and additive inverse. -/
+structure add_subgroup (G : Type*) [add_group G] extends add_submonoid G:=
+(neg_mem' {x} : x ∈ carrier → -x ∈ carrier)
+
+attribute [to_additive add_subgroup] subgroup
+
+/-- Create a bundled subgroup from a set `s` and `[is_subroup s]`. -/
+@[to_additive "Create a bundled additive subgroup from a set `s` and `[is_add_subgroup s]`."]
+def subgroup.of (s : set G) [h : is_subgroup s] : subgroup G :=
+{ carrier := s,
+  one_mem' := h.1.1,
+  mul_mem' := h.1.2,
+  inv_mem' := h.2 }
+
+/-- Map from subgroups of group `G` to `add_subgroup`s of `additive G`. -/
+def subgroup.to_add_subgroup {G : Type*} [group G] (H : subgroup G) :
+  add_subgroup (additive G) :=
+{ inv_mem' := H.inv_mem',
+  .. submonoid.to_add_submonoid H.to_submonoid}
+
+/-- Map from `add_subgroup`s of `additive G` to subgroups of `G`. -/
+def subgroup.of_add_subgroup {G : Type*} [group G] (H : add_subgroup (additive G)) :
+  subgroup G :=
+{ inv_mem' := H.inv_mem',
+  .. submonoid.of_add_submonoid H.to_add_submonoid}
+
+/-- Map from `add_subgroup`s of `add_group G` to subgroups of `multiplicative G`. -/
+def add_subgroup.to_subgroup {G : Type*} [add_group G] (H : add_subgroup G) :
+  subgroup (multiplicative G) :=
+{ inv_mem' := H.inv_mem',
+  .. add_submonoid.to_submonoid H.to_add_submonoid}
+
+/-- Map from subgroups of `multiplicative G` to `add_subgroup`s of `add_group G`. -/
+def add_subgroup.of_subgroup {G : Type*} [add_group G] (H : subgroup (multiplicative G)) :
+  add_subgroup G :=
+{ inv_mem' := H.inv_mem',
+  .. add_submonoid.of_submonoid H.to_submonoid }
+
+/-- Subgroups of group `G` are isomorphic to additive subgroups of `additive G`. -/
+def subgroup.add_subgroup_equiv (G : Type*) [group G] :
+subgroup G ≃ add_subgroup (additive G) :=
+{ to_fun := subgroup.to_add_subgroup,
+  inv_fun := subgroup.of_add_subgroup,
+  left_inv := λ x, by cases x; cases x__to_submonoid; refl,
+  right_inv := λ x, by cases x; cases x__to_add_submonoid; refl }
+
+namespace subgroup
+
+@[to_additive]
+instance : has_coe (subgroup G) (set G) := { coe := λ K, K.to_submonoid.carrier }
+
+@[to_additive]
+instance : has_mem G (subgroup G) := ⟨λ m K, m ∈ (K : set G)⟩
+
+@[simp, to_additive]
+lemma mem_coe {K : subgroup G} [g : G] : g ∈ (K : set G) ↔ g ∈ K := iff.rfl
+
+@[simp, squash_cast, to_additive, nolint simp_nf]
+lemma coe_coe (K : subgroup G) : ↥(K : set G) = K := rfl
+
+attribute [squash_cast, nolint simp_nf] add_subgroup.coe_coe
```

Comment:
Please merge `master` and replace all `*_cast` with `norm_cast`. Now it can figure out the type of a `*_cast` lemma automatically.

Reply:
I'm not sure what do you mean by "replace all `*_cast` with `norm_cast`". Do I replace `squash_cast` with `norm_cast`?

Reply:
Yes, and also change any `elim_cast` and `move_cast` to `norm_cast`. See #2407.

Reply:
It looks like removing the `nolint simp_nf` attributes caused [the linter to fail on these again](https://github.com/leanprover-community/mathlib/runs/605669371#step:14:68).

Reply:
I guess I should add it back. I saw that it was removed for `submonoid.lean` so I expected it would work for my case but I suppose not :(

Reply:
Yeah, I'm not sure what's going on here. @gebner any thoughts?